Lesson 3 HW

grrrvahrrr opened the pull request:
Еще поигрался с Гитом, надеюсь все работает)

Написал калькулятор на базе своего кода из предыдущего ДЗ.
В это задании потренировался работать с функциями и с проверкой на ошибки. Очень хотелось в своих функциях записывать err не из Scanf(что вызывает свои проблемы), а проверить соответствует ли тип введенной переменной float64 и если нет, то выдавать ошибку через fmt.Errorf как было на лекции. Но углубившись в вопрос, я понял что проверка на тип производится через интерфейс и так как мы его еще не проходили, решил с этим отложить) Может быть есть какая-то другая проверка на тип? Сработал бы способ с проверкой на тип?

Вывод простых чисел включил в калькулятор. С математикой у меня не супер, но так как похожее задание было в курсе JS тело решения взял оттуда, но переписал его на Го для тренировки написания циклов.

grrrvahrrr committed:
playing with github

grrrvahrrr committed:
Lesson 3 HW

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,152 @@
+	Multiplication = "*"     //multiplying numbers
+	Division       = "/"     //divising numbers
+	Squared        = "sqr"   //squared numbers
+	SqrRoot        = "root"  //square root
```
`sqrt` - принятое сокращение

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,152 @@
+var number1, number2 float64
+var err error
+
+func twoNumberOperation() (float64, float64, error) {
```
Для читаемости, может только почистить от излишних пустых строк (перед fmt.Println() первым, перед err !=nil, после return

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,152 @@
+	return number1, nil
+}
+
+func main() {
```
тут так же, слишком много лишних "пустых строк". уменьшает читаемость кода

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,152 @@
+		os.Exit(1)
+	}
+
+	return number1, number2, nil
```
функция возвращает глобальные переменные - не хорошо. Переделайте

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,152 @@
+		for i = 1; i < mbPrime; i++ {
+			isPrime = true
+
+			for j = 2; j < i; j++ {
```
подумайте, как можно оптимизировать цикл. от 2 до i это слишком большой диапазон

grrrvahrrr replied:
Сделал новый коммит с изменениями. Раз это приходит на почту, то вот сразу ссылка)
https://github.com/grrrvahrrr/GolangLvl1/pull/3/commits/249085de0d83e56af5caca09d4d0500f9399cf4b

goarchitecture replied:
подумайте ещё. 
и зачем вы добавляете отдельное условие `if j > mbPrime /2` если у вас есть объявления цикла, где вы ограничиваете его диапазоны

goarchitecture reviewed (changes requested):
Отлично (авансом). Всё хорошо, кроме "глобальных переменных". В комментариях описал

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,152 @@
+	Prime          = "prime" //Get prime numbers upto this point
+)
+
+var number1, number2 float64
```
не нужны

grrrvahrrr committed:
Lesson 3 HW Update

Убрал глобальные переменные, пустые строки.
Подумал над оптимизацией кода на простые числа - вроде получилось сократить проверку вдвое, но с математикой все еще не очень, учусь)

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,148 @@
+	_, err = fmt.Scanf("%f %f", &number1, &number2)
+	if err != nil {
+		fmt.Println(`Try again, something went wrong..`)
+		os.Exit(1)
```
функция не должна решать что делать при ошибке: просто возвращайте её. А уже тот кто вызывает функцию - решит что делать с ошибкой

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,148 @@
+	_, err = fmt.Scanf("%f", &number1)
+	if err != nil {
+		fmt.Println(`Try again, something went wrong..`)
+		os.Exit(1)
```
тоже самое - не нагружайте функцию - она просто возвращает ошибку, а не принимает действия

goarchitecture reviewed (approved):
добавил ещё код-ревью

grrrvahrrr committed:
Lesson 3 HW Update 2

Наконец разобрался с функциями, вроде понял про простые числа)

grrrvahrrr committed:
Bug hunting

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,158 @@
+
+		_, err = fmt.Scanf("%d", &mbPrime)
+		if err != nil {
+			fmt.Println(`Try again, something went wrong..`)
```
Something went wrong - не понятно, что именно went wrong. юзер должен знать в чем причина и может ли он повлиять (здесь и везде - где так же обрабатывается ошибка)

goarchitecture commented on Lesson3/main.go:
```diff
@@ -0,0 +1,158 @@
+			num := 3
+			for num <= mbPrime {
+				isPrime = true
+				for i := 2; i <= int(math.Sqrt(float64(num))); i++ {
```
а ещё можно упростить шаг цикла. Если мы првоерили что не делится на 2, то все остальные четные "неделения" - не актуальны (на 4, на 6, на 8 и т.д.)

goarchitecture reviewed (approved):
Добавил пару незначительных комментариев.

в общем всё принято.
